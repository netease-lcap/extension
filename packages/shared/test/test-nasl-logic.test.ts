import { expect, test } from 'vitest';
import * as parser from '@babel/parser';
import traverse from '@babel/traverse';
import transformFunc2NaslLogic from '../src/transforms/transform-func2nasl-logic';

const code = `
/**
 * @NaslLogic
 * @title 过滤列表
 * @desc 泛型示例-过滤列表中的 null 数据
 * @param list 列表
 * @returns 列表
 */
export function filterNull<T>(list: nasl.collection.List<T>): nasl.collection.List<T> {
    return list.filter(n => !!n);
}

/**
 * @NaslLogic
 * @title 过滤列表
 * @desc 泛型示例-过滤列表中的 null 数据
 * @param list 列表
 * @returns 列表
 */
export function filter<T>(list: T[], callback: (n: T) => boolean): nasl.collection.List<T> {
    return list.filter(n => !!n);
}

/**
 * @NaslLogic
 * @title 过滤列表异步
 * @desc 泛型示例-过滤列表中的 null 数据
 * @param list 列表
 * @returns 列表
 */
export async function filterAsync<T>(list: T[], callback: (n: T) => boolean): Promise<nasl.collection.List<T>> {
    return list.filter(n => !!n);
}
`;
test('test nasl logic', () => {
  const logics: any[] = [];
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['typescript', 'jsx'],
  });

  traverse(ast, {
    ExportNamedDeclaration(p) {
      const logic = transformFunc2NaslLogic(p.node);
      if (logic) {
        logics.push(logic);
      }
    },
  });

  expect(logics).toMatchSnapshot();
});
