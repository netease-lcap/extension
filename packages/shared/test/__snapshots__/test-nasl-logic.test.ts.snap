// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test nasl logic 1`] = `
[
  {
    "body": [],
    "concept": "Logic",
    "cron": "",
    "description": "泛型示例-过滤列表中的 null 数据",
    "name": "filterNull",
    "overridable": false,
    "params": [
      {
        "concept": "Param",
        "description": "列表",
        "name": "list",
        "spread": false,
        "typeAnnotation": {
          "concept": "TypeAnnotation",
          "ruleMap": {},
          "typeArguments": [
            {
              "concept": "TypeAnnotation",
              "ruleMap": {},
              "typeArguments": [],
              "typeKind": "typeParam",
              "typeName": "T",
              "typeNamespace": "",
            },
          ],
          "typeKind": "generic",
          "typeName": "List",
          "typeNamespace": "nasl.collection",
        },
      },
    ],
    "playground": [],
    "returns": [
      {
        "concept": "Return",
        "description": "列表",
        "name": "result",
        "typeAnnotation": {
          "concept": "TypeAnnotation",
          "ruleMap": {},
          "typeArguments": [
            {
              "concept": "TypeAnnotation",
              "ruleMap": {},
              "typeArguments": [],
              "typeKind": "typeParam",
              "typeName": "T",
              "typeNamespace": "",
            },
          ],
          "typeKind": "generic",
          "typeName": "List",
          "typeNamespace": "nasl.collection",
        },
      },
    ],
    "sync": true,
    "title": "过滤列表",
    "triggerType": "",
    "type": "",
    "typeParams": [
      {
        "concept": "TypeParam",
        "displayName": "",
        "name": "T",
      },
    ],
    "variables": [],
  },
  {
    "body": [],
    "concept": "Logic",
    "cron": "",
    "description": "泛型示例-过滤列表中的 null 数据",
    "name": "filter",
    "overridable": false,
    "params": [
      {
        "concept": "Param",
        "description": "列表",
        "name": "list",
        "spread": false,
        "typeAnnotation": {
          "concept": "TypeAnnotation",
          "ruleMap": {},
          "typeArguments": [
            {
              "concept": "TypeAnnotation",
              "ruleMap": {},
              "typeArguments": [],
              "typeKind": "typeParam",
              "typeName": "T",
              "typeNamespace": "",
            },
          ],
          "typeKind": "generic",
          "typeName": "List",
          "typeNamespace": "nasl.collection",
        },
      },
      {
        "concept": "Param",
        "description": undefined,
        "name": "callback",
        "spread": false,
        "typeAnnotation": {
          "concept": "TypeAnnotation",
          "returnType": [
            {
              "concept": "TypeAnnotation",
              "ruleMap": {},
              "typeKind": "primitive",
              "typeName": "Boolean",
              "typeNamespace": "nasl.core",
            },
          ],
          "ruleMap": {},
          "sync": true,
          "typeArguments": [
            {
              "concept": "TypeAnnotation",
              "ruleMap": {},
              "typeArguments": [],
              "typeKind": "typeParam",
              "typeName": "T",
              "typeNamespace": "",
            },
          ],
          "typeKind": "function",
          "typeName": "",
          "typeNamespace": "",
        },
      },
    ],
    "playground": [],
    "returns": [
      {
        "concept": "Return",
        "description": "列表",
        "name": "result",
        "typeAnnotation": {
          "concept": "TypeAnnotation",
          "ruleMap": {},
          "typeArguments": [
            {
              "concept": "TypeAnnotation",
              "ruleMap": {},
              "typeArguments": [],
              "typeKind": "typeParam",
              "typeName": "T",
              "typeNamespace": "",
            },
          ],
          "typeKind": "generic",
          "typeName": "List",
          "typeNamespace": "nasl.collection",
        },
      },
    ],
    "sync": true,
    "title": "过滤列表",
    "triggerType": "",
    "type": "",
    "typeParams": [
      {
        "concept": "TypeParam",
        "displayName": "",
        "name": "T",
      },
    ],
    "variables": [],
  },
  {
    "body": [],
    "concept": "Logic",
    "cron": "",
    "description": "泛型示例-过滤列表中的 null 数据",
    "name": "filterAsync",
    "overridable": false,
    "params": [
      {
        "concept": "Param",
        "description": "列表",
        "name": "list",
        "spread": false,
        "typeAnnotation": {
          "concept": "TypeAnnotation",
          "ruleMap": {},
          "typeArguments": [
            {
              "concept": "TypeAnnotation",
              "ruleMap": {},
              "typeArguments": [],
              "typeKind": "typeParam",
              "typeName": "T",
              "typeNamespace": "",
            },
          ],
          "typeKind": "generic",
          "typeName": "List",
          "typeNamespace": "nasl.collection",
        },
      },
      {
        "concept": "Param",
        "description": undefined,
        "name": "callback",
        "spread": false,
        "typeAnnotation": {
          "concept": "TypeAnnotation",
          "returnType": [
            {
              "concept": "TypeAnnotation",
              "ruleMap": {},
              "typeKind": "primitive",
              "typeName": "Boolean",
              "typeNamespace": "nasl.core",
            },
          ],
          "ruleMap": {},
          "sync": true,
          "typeArguments": [
            {
              "concept": "TypeAnnotation",
              "ruleMap": {},
              "typeArguments": [],
              "typeKind": "typeParam",
              "typeName": "T",
              "typeNamespace": "",
            },
          ],
          "typeKind": "function",
          "typeName": "",
          "typeNamespace": "",
        },
      },
    ],
    "playground": [],
    "returns": [
      {
        "concept": "Return",
        "description": "列表",
        "name": "result",
        "typeAnnotation": {
          "concept": "TypeAnnotation",
          "ruleMap": {},
          "typeArguments": [
            {
              "concept": "TypeAnnotation",
              "ruleMap": {},
              "typeArguments": [],
              "typeKind": "typeParam",
              "typeName": "T",
              "typeNamespace": "",
            },
          ],
          "typeKind": "generic",
          "typeName": "List",
          "typeNamespace": "nasl.collection",
        },
      },
    ],
    "sync": false,
    "title": "过滤列表异步",
    "triggerType": "",
    "type": "",
    "typeParams": [
      {
        "concept": "TypeParam",
        "displayName": "",
        "name": "T",
      },
    ],
    "variables": [],
  },
]
`;
